src/browser/modules/D3Visualization/lib/visualization/utils/textMeasurement.ts

```diff
@@ -17,43 +17,43 @@
  * You should have received a copy of the GNU General Public License
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
-import d3 from 'd3'
 
-const measureUsingCanvas = function(text: any, font: any) {
-  // @ts-expect-error ts-migrate(2683) FIXME: 'this' implicitly has type 'any' because it does n... Remove this comment to see the full error message
-  const canvasSelection = d3.select('canvas#textMeasurementCanvas').data([this])
-  canvasSelection
-    .enter()
-    .append('canvas')
-    .attr('id', 'textMeasurementCanvas')
-    .style('display', 'none')
-
-  const canvas = canvasSelection.node() as HTMLCanvasElement
-  const context = canvas.getContext('2d') as CanvasRenderingContext2D
+const measureTextWidthByCanvas = (
+  text: string,
+  font: string,
+  context: CanvasRenderingContext2D
+): number => {
   context.font = font
   return context.measureText(text).width
 }
 
-const cache = function() {
-  const cacheSize = 10000
-  const map: any = {}
-  const list: any = []
-  return (key: any, calc: any) => {
-    const cached = map[key]
+const cacheTextWidth = function() {
+  const CATCH_SIZE = 100000
+  const textMeasureMap: { [key: string]: number } = {}
+  const lruKeyList: string[] = []
+  return (key: string, calculate: () => number) => {
+    const cached = textMeasureMap[key]
     if (cached) {
       return cached
     } else {
-      const result = calc()
-      if (list.length > cacheSize) {
-        delete map[list.splice(0, 1)]
-        list.push(key)
+      const result = calculate()
+      if (lruKeyList.length > CATCH_SIZE) {
+        delete textMeasureMap[lruKeyList.splice(0, 1).toString()]
+        lruKeyList.push(key)
       }
-      return (map[key] = result)
+      return (textMeasureMap[key] = result)
     }
   }
 }
 
-export default function(text: any, fontFamily: any, fontSize: any) {
+export default function(
+  text: string,
+  fontFamily: string,
+  fontSize: number,
+  canvas2DContext: CanvasRenderingContext2D
+): number {
   const font = `normal normal normal ${fontSize}px/normal ${fontFamily}`
-  return cache()(text + font, () => measureUsingCanvas(text, font))
+  return cacheTextWidth()(`[${font}][${text}]`, () =>
```

Comment:
We talked about taking the shortcut of using width of the widest available letter, did you decide not to pursuit that option or did you just not want to include it in this work?

Reply:
@OskarDamkjaer - I need to check with @gregoryking if we have scenarios with different languages. If it only contains English, other latin languages and numbers, we can simply optimise this method. Otherwise, we need more investigation if it's doable for multi-languages, i.e. Arabic. Chinese, Japanese, Thai, etc. Potential problems are:
- A letter/character has different widths in different fonts (since we don't allow users to change font, it should be fine).
- Even one font can be rendered with slightly different widths in different browsers (that's why we need to create canvas to calculate the width under the runtime of certain browser).
- Symbols have different widths (including symbols' padding). So, we need to consider this case.

So, I think better **_not_** to include this work in this PR. But I'll keep it on my radar when I get back from @gregoryking 💪 .

Reply:
@OskarDamkjaer - from this Trello card, we also need to consider emoji: https://trello.com/c/DcFUZlDU.

Reply:
Sounds good, thanks!